### PR TITLE
Fix ancestors_of exception for tables with >=2000 snapshots

### PR DESCRIPTION
**Pull request by ndrluis**
When I was running some tests in production, I discovered this error.

### CODE REVIEW

**Comment by syun64** [comment]
Hi @ndrluis - thank you for raising this. This looks like an issue we'd definitely want to fix.

Just to help with reviewing this PR, and for future records, would you be able to share the stack trace of the exception you are running into in this edge case?

**Comment by ndrluis** [comment]
Hi @syun64, this is the stacktrace

```
Traceback (most recent call last):
  File "/Users/magus/Workspace/oss/iceberg-python/test_view.py", line 18, in <module>
    print(list(ancestors_of(t.current_snapshot(), t.metadata)))
  File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 428, in ancestors_of
    yield from ancestors_of(parent, table_metadata)
  File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 428, in ancestors_of
    yield from ancestors_of(parent, table_metadata)
  File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 428, in ancestors_of
    yield from ancestors_of(parent, table_metadata)
  [Previous line repeated 992 more times]
  File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 427, in ancestors_of
    if parent := table_metadata.snapshot_by_id(current_snapshot.parent_snapshot_id):
  File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/metadata.py", line 231, in snapshot_by_id
    return next((snapshot for snapshot in self.snapshots if snapshot.snapshot_id == snapshot_id), None)
  File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/metadata.py", line 231, in <genexpr>
    return next((snapshot for snapshot in self.snapshots if snapshot.snapshot_id == snapshot_id), None)
RecursionError: maximum recursion depth exceeded in comparison
```

File: pyiceberg/table/snapshots.py [hunk 1]
```diff
@@ -421,8 +421,16 @@ def set_when_positive(properties: Dict[str, str], num: int, property_name: str)
 
 def ancestors_of(current_snapshot: Optional[Snapshot], table_metadata: TableMetadata) -> Iterable[Snapshot]:
     """Get the ancestors of and including the given snapshot."""
-    if current_snapshot:
-        yield current_snapshot
-        if current_snapshot.parent_snapshot_id is not None:
-            if parent := table_metadata.snapshot_by_id(current_snapshot.parent_snapshot_id):
-                yield from ancestors_of(parent, table_metadata)
+    seen = set()
```
Comment (Fokko) [inline pyiceberg/table/snapshots.py:424]: Out of curiosity, why do we need to track the seen ones? The metadata should be a tree, and not a graph (no cycles)

**Comment by syun64** [comment]
> Hi @syun64, this is the stacktrace
> 
> ```
> Traceback (most recent call last):
>   File "/Users/magus/Workspace/oss/iceberg-python/test_view.py", line 18, in <module>
>     print(list(ancestors_of(t.current_snapshot(), t.metadata)))
>   File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 428, in ancestors_of
>     yield from ancestors_of(parent, table_metadata)
>   File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 428, in ancestors_of
>     yield from ancestors_of(parent, table_metadata)
>   File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 428, in ancestors_of
>     yield from ancestors_of(parent, table_metadata)
>   [Previous line repeated 992 more times]
>   File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/snapshots.py", line 427, in ancestors_of
>     if parent := table_metadata.snapshot_by_id(current_snapshot.parent_snapshot_id):
>   File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/metadata.py", line 231, in snapshot_by_id
>     return next((snapshot for snapshot in self.snapshots if snapshot.snapshot_id == snapshot_id), None)
>   File "/Users/magus/Workspace/oss/iceberg-python/pyiceberg/table/metadata.py", line 231, in <genexpr>
>     return next((snapshot for snapshot in self.snapshots if snapshot.snapshot_id == snapshot_id), None)
> RecursionError: maximum recursion depth exceeded in comparison
> ```

Hey @ndrluis  thanks for sharing the stacktrace. If the issue is with the recursion depth, could we try writing this in a way that doesn't require recursion?

i.e.

```
snapshot = current_snapshot
while snapshot is not None:
    yield snapshot
    snapshot = table_metadata.snapshot_by_id(current_snapshot.parent_snapshot_id)
```

File: pyiceberg/table/snapshots.py (hunk 1 above)
Comment (ndrluis) [reply pyiceberg/table/snapshots.py]: You are right, this was added because I thought there was an infinite loop occurring.

**Comment by ndrluis** [comment]
@syun64 Thank you for the review. I made the changes as you asked and also made a slight change due to a mypy error regarding the snapshot_by_id type `Optional[int]`.

**Review by syun64 — APPROVED** [review]
LGTM

**Review by Fokko — APPROVED** [review]
@ndrluis Thanks, looking good 👍 @syun64 appreciate the review 🙌

**Pull request merged**